src/ai/defaultai.cc

```diff
@@ -6237,13 +6367,19 @@ void DefaultAI::consider_productionsite_influence(BuildableField& field,
 	for (const Widelands::DescriptionIndex& i : bo.ware_outputs) {
 		++field.producers_nearby.at(i);
 	}
-	if (bo.has_collected_map_resource()) {
-		++field.collecting_producers_nearby.at(bo.get_collected_map_resource());
+	for (const auto& supported_building : bo.supported_producers) {
+		// Using std::map syntax to create entry on demand.
+		++field.supported_producers_nearby[supported_building.first];
 	}
 
-	if (!bo.production_hints.empty()) {
-		for (auto ph : bo.production_hints) {
-			++field.supporters_nearby.at(ph);
+	if (bo.type == BuildingObserver::Type::kProductionsite) {
```

Comment:
this should have been already a prerequisite for being here.

Reply:
Should but isn't. I'll move the check to the calling code:
```

if (player_immovable->descr().type() == Widelands::MapObjectType::PRODUCTIONSITE) {
  BuildingObserver& bo = get_building_observer(player_immovable->descr().name().c_str());
  consider_productionsite_influence(field, imm_found.coords, bo);
} else if (upcast(Widelands::ConstructionSite const, constructionsite, player_immovable)) {
  const Widelands::BuildingDescr& target_descr = constructionsite->building();
  // Check goes here
  BuildingObserver& bo = get_building_observer(target_descr.name().c_str());
  consider_productionsite_influence(field, imm_found.coords, bo);
}
```